src/main/host/syscall/socket.c

```diff
@@ -361,6 +384,67 @@ static int _syscallhandler_getSocketOptHelper(SysCallHandler* sys, Socket* sock,
     }
 }
 
+static int _syscallhandler_setTCPOptHelper(SysCallHandler* sys, TCP* tcp, int optname,
+                                           PluginPtr optvalPtr, socklen_t optlen) {
+    switch (optname) {
+        case TCP_NODELAY: {
+            /* Shadow doesn't support nagle's algorithm, so shadow always behaves as if TCP_NODELAY
+             * is enabled. Some programs will fail if setsockopt(fd, SOL_TCP, TCP_NODELAY, &1,
+             * sizeof(int)) returns an error, so we treat this as a no-op for compatibility.
+             */
+            if (optlen < sizeof(int)) {
+                return -EINVAL;
+            }
+
+            int enable = 0;
+            int errcode = process_readPtr(sys->process, &enable, optvalPtr, sizeof(int));
+            if (errcode != 0) {
+                return errcode;
+            }
+
+            if (enable) {
+                // wants to enable TCP_NODELAY
+                debug("Ignoring TCP_NODELAY");
+            } else {
+                // wants to disable TCP_NODELAY
+                warning("Cannot disable TCP_NODELAY since shadow does not implement "
+                        "Nagle's algorithm.");
+                return -ENOPROTOOPT;
+            }
+
+            return 0;
+        }
+        case TCP_CONGESTION: {
+            // the value of TCP_CA_NAME_MAX in linux
+            const int CONG_NAME_MAX = 16;
+
+            char name[CONG_NAME_MAX];
+            optlen = MIN(optlen, CONG_NAME_MAX);
+
+            int errcode = process_readPtr(sys->process, name, optvalPtr, optlen);
+            if (errcode != 0) {
+                return errcode;
+            }
+
+            if (optlen < 4 || strncmp(name, "reno", optlen) != 0) {
+                warning("Shadow sockets only support 'reno' for TCP_CONGESTION");
+                return -ENOENT;
+            }
+
+            // shadow doesn't support other congestion types, so do nothing
+            const char* current_name = tcp_cong(tcp)->hooks->tcp_cong_name_str();
+            utility_assert(current_name != NULL && strcmp(current_name, "reno") == 0);
```

Comment:
Could we expose this constant in `tcp_cong.h` rather than repeating it here? e.g. by exposing `tcp_cong_reno_name_str_()`?

Reply:
Added this as a constant so the string is only in one place.